### PR TITLE
Add tutorial for integration testing using virtual machines

### LINKED ISSUE #111: NixOS tests

### PR DESCRIPTION
**Pull request by domenkozar**
Fixes #111

[rendered](https://fab33bb8.nix-dev.pages.dev/tutorials/integration-testing-using-virtual-machines)

cc @zimbatm @zupo @NorfairKing

### CODE REVIEW

File: source/tutorials/integration-testing-using-virtual-machines.rst [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    webRole       = "web_anon";
+    postgrestPort = 3000;
+
+    # NixOS module shared between server and client
+    sharedModule = {
+      # don't assume graphical interface is possible
+      virtualisation.graphics = false;
+    };
+
+  in import "${nixpkgs}/nixos/tests/make-test-python.nix" ({ pkgs, ...}: {
```
Comment (tfc) [inline source/tutorials/integration-testing-using-virtual-machines.rst:68]: you can use `pkgs.nixosTest` as a function to call

File: source/tutorials/integration-testing-using-virtual-machines.rst [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+      ]
+
+      actual = json.loads(
+          client.succeed(
+              "${pkgs.curl}/bin/curl http://server:${toString postgrestPort}/${table}"
+          )
+      )
+
+      if expected != actual:
+          sys.exit(1)
```
Comment (tfc) [inline source/tutorials/integration-testing-using-virtual-machines.rst:158]: does anything speak against this?

```python
assert expected == actual, "table query returns expected content"
```

File: source/tutorials/integration-testing-using-virtual-machines.rst [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+
+Next steps
+----------
+
+- Running integration tests on CI requires hardware acceleration, which many CIs do not support. 
+  To run integration tests on :ref:`GitHub Actions <github-actions>` see 
+  `how to disable hardware acceleration <https://github.com/cachix/install-nix-action#how-can-i-run-nixos-tests>`_.
+
+- NixOS comes with a large set of tests that serve also as educational examples, a good inspiration is `Matrix bridging with an IRC <https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/matrix-appservice-irc.nix>`_.
```
Comment (tfc) [inline source/tutorials/integration-testing-using-virtual-machines.rst:228]: my personal favorite example is the bittorrent test. i can also provide a nice picture that shows what kind of complexity the test maintains with few lines that i would share if you are interested.

Screenshot from 2021-06-03 14-18-08

File: source/tutorials/integration-testing-using-virtual-machines.rst [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    table         = "todos";
+    username      = "authenticator";
+    password      = "mysecretpassword";
+    webRole       = "web_anon";
+    postgrestPort = 3000;
+
+    # NixOS module shared between server and client
+    sharedModule = {
+      # don't assume graphical interface is possible
+      virtualisation.graphics = false;
```
Comment (NorfairKing) [inline source/tutorials/integration-testing-using-virtual-machines.rst:65]: Why is this necessary? Does it speed up the test? If so, I should add that to my own tests.

Comment (domenkozar) [reply source/tutorials/integration-testing-using-virtual-machines.rst:65]: It makes sure the tests works on CI where there's no $DISPLAY.

File: source/tutorials/integration-testing-using-virtual-machines.rst (hunk 3 above)
Comment (domenkozar) [reply source/tutorials/integration-testing-using-virtual-machines.rst:228]: That one is pretty huge, I'd like to give smaller examples that demonstrate the power of what it can do.

File: source/tutorials/integration-testing-using-virtual-machines.rst [hunk 5]
```diff
@@ -0,0 +1,228 @@
+Integration testing using Virtual Machines
+==========================================
+
+One of the most powerful features in the Nix ecosystem is **the ability
+to provide a set of declarative NixOS configurations and use a simple
+Python interface** to interact with them using `QEMU <https://www.qemu.org/>`_
+as the backend.
+
+They are also known as **NixOS tests**.
```
Comment (Patryk27) [inline source/tutorials/integration-testing-using-virtual-machines.rst:9]: Maybe it'd be worth pointing out that those tests don't require using NixOS as the host system?

Something in terms of:

> Those tests are widely used to ensure that NixOS works as intended, so in general they are called **NixOS tests** - though they can be written and launched outside of NixOS, on any Linux machine (with MacOS support coming soon).

File: source/tutorials/integration-testing-using-virtual-machines.rst (hunk 4 above)
Comment (Patryk27) [reply source/tutorials/integration-testing-using-virtual-machines.rst:66]: This one caught me off-guard, too - how about phrasing such as this?

> since it's common for CI not to have $DISPLAY available, we have to explicitly tell the tests "please don't expect any screen available"

(this concrete wording is kinda flake-y, but hopefully will work as a sales pitch here 🙂)

File: source/tutorials/integration-testing-using-virtual-machines.rst [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+    sharedModule = {
+      # it's common for CI not to have $DISPLAY available
+      virtualisation.graphics = false;
+    };
+
+  in pkgs.nixosTest ({
+    # In case we're evaluating from macOS, pin the system
+    system = "x86_64-linux";
+
+    # Disable linting for simpler debugging
```
Comment (Patryk27) [inline source/tutorials/integration-testing-using-virtual-machines.rst:73]: Because the `testScript =` thingie this parameter refers to appears waay below, how about we either move this option below `testScript`, or change the comment a bit:

> In a few lines below, there's an option called `testScript` - it contains the test's body written in Python. By the default, before the test is executed, `pkgs.nixosTest` ensures the Python code is neatly formatted and, if it's not, prevents the test from running. This is great in the long run, but makes the development a bit harder and so for the time being we're disabling linting that Python code.

File: source/tutorials/integration-testing-using-virtual-machines.rst [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+    postgrestPort = 3000;
+
+    # NixOS module shared between server and client
+    sharedModule = {
+      # it's common for CI not to have $DISPLAY available
+      virtualisation.graphics = false;
+    };
+
+  in pkgs.nixosTest ({
+    # In case we're evaluating from macOS, pin the system
```
Comment (Patryk27) [inline source/tutorials/integration-testing-using-virtual-machines.rst:70]: How about something like this?

> NixOS tests are run inside a virtual machine, and here we specify which machine should get emulated. Currently only `x86_64-linux` is supported.

(though tbh i'm not sure if only `x86_64-linux` is supported - can it actually virtualize _arbitrary_ qemu-and-nixos-compatible machine? I've never checked it before 😄)

File: source/tutorials/integration-testing-using-virtual-machines.rst [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+
+  $ nix-build postgrest.nix
+
+You'll notice an error message if something goes wrong.
+
+In case the tests succeed, you should see at the end:
+
+.. code:: shell-session
+
+  1 out of 1 tests succeeded
```
Comment (Patryk27) [inline source/tutorials/integration-testing-using-virtual-machines.rst:187]: Hmm, in my case what's returned is:
```
<!-- logs from the VMs -->
client: must succeed: /nix/store/g8w72aanshmw5gxi02ril1qlb427rjzy-curl-7.76.1-bin/bin/curl http://server:3000/todos
client: waiting for the VM to finish booting
client: connected to guest root shell
client: (connecting took 0.00 seconds)
(0.00 seconds)
server # [   10.554648] postgrest-start[908]: 192.168.1.1 - - [03/Jun/2021:17:53:41 +0000] "GET /todos HTTP/1.1" 200 - "" "curl/7.76.1"
(0.09 seconds)
(10.93 seconds)
test script finished in 10.96s
cleaning up
killing client (pid 10)
killing server (pid 22)
(0.00 seconds)
/nix/store/bx7z3imvxxpwkkza10vb23czhw7873w2-vm-test-run-unnamed
```

**Review by Patryk27 — COMMENTED** [review]
❤️

File: source/tutorials/integration-testing-using-virtual-machines.rst [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+
+
+Developing and debugging tests
+------------------------------
+
+When developing tests or when something breaks, it's useful to interactively fiddle
+with the script or access a terminal for a machine.
+
+To interactively start a Python session with testing framework:
```
[the remark this answers is not in the document]
Comment (domenkozar) [reply source/tutorials/integration-testing-using-virtual-machines.rst:196]: I try hard to not offer two solutions to the same problem and since `nixos-run-vms` doesn't allow fiddling with the test script, it's inferior to the driver method. Does that make sense?

File: source/tutorials/integration-testing-using-virtual-machines.rst (hunk 7 above)
Comment (domenkozar) [reply source/tutorials/integration-testing-using-virtual-machines.rst:70]: `aarch64-linux` should work too I imagine.

**Comment by domenkozar** [comment]
PTAL :)

**Comment by tzarsquared** [comment]
I have read the CLA Document and I hereby sign the CLA
